Crash on generic class inside a function with a forward-reference bound

This crashes mypy:
```python
from typing import Generic, TypeVar

def test() -> None:
    T = TypeVar("T", bound="Model")

    class Query(Generic[T]):
        def filter(self, value: T) -> None:
            raise NotImplementedError

    class Model:
        pass
```

mypy (both 2.1.0 and current master) crashes on this with:

AssertionError: Must not defer during final iteration

Necessary pieces, from reduction:

- The definitions need to be inside a function.
- The TypeVar needs a forward-ref bound to a later local class: bound="Model".
- The local generic class needs to use that T in a method signature.

Similar issues: #12112, #13678, #18579.

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -1156,7 +1156,8 @@ class SemanticAnalyzer(
                         leading_type = fill_typevars(info)
                     if func.is_class or func.name == "__new__":
                         leading_type = self.class_type(leading_type)
-                    func.type = replace_implicit_first_type(functype, leading_type)
+                    if not has_placeholder(leading_type):
+                        func.type = replace_implicit_first_type(functype, leading_type)
                 elif has_self_type and isinstance(func.unanalyzed_type, CallableType):
                     if not isinstance(get_proper_type(func.unanalyzed_type.arg_types[0]), AnyType):
                         if self.is_expected_self_type(

--- a/test-data/unit/check-newsemanal.test
+++ b/test-data/unit/check-newsemanal.test
@@ -3228,3 +3228,16 @@ def deco(fn: Callable[[], T]) -> Callable[[], T]: ...
 
 @deco
 def defer() -> int: ...
+
+[case testNestedClassSelfNoPlaceholder]
+from typing import Generic, TypeVar
+
+def test() -> None:
+    T = TypeVar("T", bound="Model")
+
+    class Query(Generic[T]):
+        def filter(self, value: T) -> None:
+            ...
+
+    class Model:
+        pass

